doc/source/development/rfc/rfc101_raster_dataset_threadsafety.rst

```diff
@@ -0,0 +1,484 @@
+    real    0m2.084s
+    user    0m8.155s
+    sys     0m0.020s
+
+    vs
+
+    $ time multireadtest -thread_safe -t 4 -i 1000000 20x20.tif
+    real    0m2.387s
+    user    0m9.334s
+    sys     0m0.029s
+
+
+But on a 4096x4096 raster with a number of iterations reduced to 100. the
+thread_safe mode is slightly faster:
+
+.. code-block:: shell
+
+    $ time multireadtest -t 4 -i 100 4096x4096.tif
+    real    0m4.845s
+    user    0m18.666s
+    sys     0m0.084s
+
+    vs
+
+    $ time multireadtest -thread_safe -t 4 -i 100 4096x4096.tif
+    real    0m4.255s
+    user    0m16.370s
+    sys     0m0.096s
+
+
+A Python equivalent of multireadtest has been written. Scalability depends
+on how much Python code is executed. If relatively few long-enough calls to
+GDAL are done, scalability tends to be good due to the Python Global Interpreter
+Lock (GIL) being dropped around them. If many short calls are done, the GIL
+itself, or its repeated acquisition and release, becomes the bottleneck. This is
+no different than using a GDALDataset per thread.
+
+Documentation
+-------------
+
+Documentation for the new constant and functions will be added. The
+:ref:`multithreading` page will be updated to reflect the new capability
+introduced by this RFC.
+
+Backward compatibility
+----------------------
+
+No issue anticipated: the C and C++ API are extended.
+The C++ ABI is modified due to additions of new virtual methods.
+
+Testing
+-------
+
+Tests will be added for the new functionality, including stress tests to have
+sufficiently high confidence in the correctness of the implementation for common
+use cases.
+
+Risks
+-----
+
+Like all code related to multi-threading, the C++ language and tooling offers
+hardly any safety belt against thread-safety programming errors. So it cannot
+be excluded that the implementation suffers from  bugs in some edge scenarios,
+or in the usage of some methods of GDALDataset, GDALRasterBand and related objects
+(particularly existing non-virtual methods of those classes that could happen
+to have a non thread-safe implementation)
+
+Open questions
+--------------
+
+The question in this section needs to be solved during the discussion phase of
+the RFC.
+
+For the unusual situations where a dataset cannot be reopened and thus
+GDALCreateThreadSafeDataset() fails, should we provide an additional ``bForce``
+argument to force it to still return a dataset, where calls to the wrapped
+dataset are protected by a mutex? This would enable to always write multi-thread
+safe code, even if the access to the dataset is serialized.
+Similarly we could have a
+``std::unique_ptr<GDALRasterBand> GDALCreateThreadSafeRasterBand(GDALRasterBand* poBand, int nOpenFlags, bool bForce)``
+function that would try to use GDALCreateThreadSafeDataset() internally if it
+manages to identify the dataset to which the band belongs to, and otherwise would
+fallback to protecting calls to the wrapped band with a mutex.
```

Comment:
Due to the mutex, there would definitely be serious performance degradation if most of the time is spent in reading th dataset. The only point would be ease of implementation for users that don't want to have to implement both a multi-thread efficient code path and a serialized code path. For some algorithms, input pixel acquisition might not be that consuming regarding other processing done, so that could still be worthwhile. This really depends on use cases.